CMakeLists.txt

```diff
@@ -748,6 +748,19 @@ if(CONFIG_GEN_ISR_TABLES)
   set_property(GLOBAL APPEND PROPERTY GENERATED_KERNEL_SOURCE_FILES isr_tables.c)
 endif()
 
+# dev_handles.c is generated from ${ZEPHYR_PREBUILT_EXECUTABLE} by
+# gen_handles.py
+add_custom_command(
+  OUTPUT dev_handles.c
+  COMMAND
+  ${PYTHON_EXECUTABLE}
+  ${ZEPHYR_BASE}/scripts/gen_handles.py
+  --output-source dev_handles.c
+  --kernel $<TARGET_FILE:${ZEPHYR_PREBUILT_EXECUTABLE}>
+  DEPENDS ${ZEPHYR_PREBUILT_EXECUTABLE}
+  )
+set_property(GLOBAL APPEND PROPERTY GENERATED_KERNEL_SOURCE_FILES dev_handles.c)
```

Comment:
this actually has the side-effect that there will always be two stage linking.

Today, if `CONFIG_GEN_ISR_TABLES=n`
https://github.com/zephyrproject-rtos/zephyr/blob/201a8d29cdb2f856f7f889f9a75d3a463e3491ef/CMakeLists.txt#L716-L749

and `CONFIG_USERSPACE=n`
https://github.com/zephyrproject-rtos/zephyr/blob/201a8d29cdb2f856f7f889f9a75d3a463e3491ef/CMakeLists.txt#L756-L894

then the there is only a single linking stage, and the pre-built executable == final executable.
https://github.com/zephyrproject-rtos/zephyr/blob/201a8d29cdb2f856f7f889f9a75d3a463e3491ef/CMakeLists.txt#L1020-L1023

This PR changes that by always requiring two times linking.

@daor-oti @mped-oticon could you provide feedback on whether this approach impacts custom toolchains.
If I recall correctly, your build system disables `USERSPACE` and `GEN_ISR_TABLES` and thus you are not postprocessing the zephyr pre-built elf file.

Reply:
@tejlmand Correct, we have disabled `USERSPACE` and `GEN_ISR_TABLES` and thus run with a single stage linking. The overhead with the two stage linking would be undesirable in our case, so perhaps it could be configurable, whether or not to run `gen_handles.py`?

Reply:
> it could be configurable, whether or not to run `gen_handles.py`?

No, device dependency information is critical for power management and safe initialization order.  If you cannot tolerate a second link operation when building then I'll have to move to mapping between device handles and actual devices by runtime code rather than an O(1) lookup.

Reply:
@pabigot Ok, it was just a thought that maybe it could be disabled, but it cannot. Unless others have a problem with the extra linking step and the small increase in build time it adds, I do not think at I can justify the rework and runtime overhead for moving to runtime code, so do not change it on just my account.